src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -712,28 +704,11 @@ private TypeSymbol ErrorType(string name = "")
         /// decision when no decision appears to match. This implementation is nonrecursive to avoid
         /// overflowing the compiler's evaluation stack when compiling a large switch statement.
         /// </summary>
-        private BoundDecisionDag MakeBoundDecisionDag(SyntaxNode syntax, ArrayBuilder<StateForCase> cases)
+        private BoundDecisionDag MakeBoundDecisionDag(SyntaxNode syntax, ref TemporaryArray<StateForCase> cases)
         {
-            // A work list of DagStates whose successors need to be computed
-            var workList = ArrayBuilder<DagState>.GetInstance();
-
             // Build the state machine underlying the decision dag
-            DecisionDag decisionDag = MakeDecisionDag(cases, workList);
-
-            workList.Free();
+            var boundDecisionDag = MakeBoundDecisionDagWorker(syntax, ref cases);
 
-            // Note: It is useful for debugging the dag state table construction to set a breakpoint
-            // here and view `decisionDag.Dump()`.
-            ;
-
-            // Compute the bound decision dag corresponding to each node of decisionDag, and store
-            // it in node.Dag.
-            var defaultDecision = new BoundLeafDecisionDagNode(syntax, _defaultLabel);
-            ComputeBoundDecisionDagNodes(decisionDag, defaultDecision);
-
-            var rootDecisionDagNode = decisionDag.RootNode.Dag;
-            RoslynDebug.Assert(rootDecisionDagNode != null);
-            var boundDecisionDag = new BoundDecisionDag(rootDecisionDagNode.Syntax, rootDecisionDagNode);
```

Comment:
moved this all into the single helper.  It made cleanup of dagStates much easier (All in one lcoation) instead of having it be spread out (i.e. created in one method, and freed in another).